### PR TITLE
improve tag image

### PR DESCRIPTION
**Pull request by jimmyxian**
Improve tag image so that it can refresh image in cluster.
Signed-off-by: Xian Chaobo <xianchaobo@huawei.com>

### CODE REVIEW

**Comment by jimmyxian** [comment]
ping @aluzzardi @vieux @abronan

**Comment by vieux** [comment]
I don't see the improvement here.

**Comment by jimmyxian** [comment]
@vieux 
I think `Engine.RefreshImages()` should not be a public function. It should not be called in docker api. If all the three API (`docker tag\docker commit\docker build`)  do as the following way, it will be  become a private function.
```
docker api ---> cluster api ---> engine api  ---> dockerclient api
```
I don't known if it's meaningful, waiting for your reply. Tks
/cc @aluzzardi @abronan

**Comment by vieux** [comment]
@jimmyxian I agree, it's an improvment regarding the quality of the code, not the feature.

LGTM

File: cluster/swarm/cluster.go [hunk 1]
```diff
@@ -410,3 +410,17 @@ func (c *Cluster) RenameContainer(container *cluster.Container, newName string)
...
+
+// TagImage tag an image
+func (c *Cluster) TagImage(IDOrName string, repo string, tag string, force bool) error {
+	c.RLock()
+	defer c.RUnlock()
+
+	// check image
+	image := c.Image(IDOrName)
+	if image == nil {
+		return fmt.Errorf("could not find image: no such id: %s", IDOrName)
```
Comment (vieux) [inline cluster/swarm/cluster.go]: can you please use `fmt.Errorf("No such image %s", IDOrName)`

Comment (jimmyxian) [inline cluster/swarm/cluster.go]: OK

**Comment by jimmyxian** [comment]
ping @vieux 
Modified by the comment. check again, thks

**Comment by aluzzardi** [comment]
I think I'm the one who suggested this, but now I'm having second thoughts.

I'm worried that the more we don't rely on proxy and use our own implementation, the more things will break since we need to re-implement them every time.

I realized this with `ps`: We don't proxy `ps` (we can't), and realized `--filter` doesn't work and since docker 1.6 they added `--filter label=k=v` which we needed to add.

I don't know if this is the path we want to take.

@vieux ?

**Comment by vieux** [comment]
@aluzzardi I have the same concerns.

**Comment by aluzzardi** [comment]
Perhaps we should address these in 0.4 with #748

**Comment by jimmyxian** [comment]
@aluzzardi Agree, using proxy can reduce a lot of work in our swarm. If we implement in swarm, Docker API changes, we also have to change it in swarm.
@aluzzardi @vieux  Shall I close this one?

**Comment by aluzzardi** [comment]
I think so yeah

**Pull request closed without merging**